mrjob/dataproc.py

```diff
@@ -0,0 +1,873 @@
+
+from mrjob.logs.counters import _pick_counters
+from mrjob.fs.gcs import is_gcs_uri
+from mrjob.parse import is_uri
+
+from mrjob.runner import MRJobRunner
+from mrjob.runner import RunnerOptionStore
+from mrjob.setup import BootstrapWorkingDirManager
+from mrjob.setup import UploadDirManager
+from mrjob.setup import parse_setup_cmd
+from mrjob.step import StepFailedException
+from mrjob.util import cmd_line
+from mrjob.util import random_identifier
+
+log = logging.getLogger(__name__)
+
+_DEFAULT_GCP_REGION = 'global'
+_HADOOP_STREAMING_JAR_URI = 'file:///usr/lib/hadoop-mapreduce/hadoop-streaming.jar'
+
+# TODO - mtai @ davidmarin - Example invocation options for mrjob.py
+#   --runner dataproc --gcp-project google.com:pm-hackathon --gcp-zone us-central1-b --gcp-cluster vbp04 --no-output gs://boulder-input-data/*.txt --output gs://boulder-input-data/mrjob-boulder
+
+# TODO - mtai @ davidmarin - Re-implement SSH tunneling? - Content pulled from job run output
+# The url to track the job: http://default-taim-m:8088/proxy/application_1456679373146_0002/
+
+# TODO - mtai @ davidmarin - Re-implement logs parsing?  See dataproc metainfo and driver output - ask Dennis Huo for more details
+# 'gs://dataproc-801485be-0997-40e7-84a7-00926031747c-us/google-cloud-dataproc-metainfo/8b76d95e-ebdc-4b81-896d-b2c5009b3560/jobs/mr_most_used_word-taim-20160228-172000-034993---Step-2-of-2/driveroutput'
+
+# TODO - mtai @ davidmarin - We should re-instance the bucket creation logic used in EMR
+_DEFAULT_GCS_BUCKET = 'boulder-input-data'
+
+_GCP_CLUSTER_NAME_REGEX = '(?:[a-z](?:[-a-z0-9]{0,53}[a-z0-9])?).'
+
+def _cleanse_gcp_job_id(job_id):
+    return re.sub('[^a-zA-Z0-9_\-]', '-', job_id)
+
+def _cleanse_gcp_cluster_name(cluster_name):
+    pass
+
+
+class DataprocRunnerOptionStore(RunnerOptionStore):
+    ALLOWED_KEYS = RunnerOptionStore.ALLOWED_KEYS.union(set([
+        'gcp_project',
+        'gcp_region',
+        'gcp_zone',
+        'gcp_cluster',
+
+        'gce_machine_type',
+        'gce_num_instances',
+
+        'gcs_sync_wait_time',
+        'gcs_tmp_dir',
+
+        'check_api_status_every',
+        'bootstrap',
+        'bootstrap_python'
+
+    ]))
+
+    COMBINERS = combine_dicts(RunnerOptionStore.COMBINERS, {
+        'bootstrap': combine_lists,
+        'gcs_tmp_dir': combine_paths,
+    })
+
+    DEPRECATED_ALIASES = combine_dicts(RunnerOptionStore.DEPRECATED_ALIASES, {
+    })
+
+    def __init__(self, alias, opts, conf_paths):
+        super(DataprocRunnerOptionStore, self).__init__(alias, opts, conf_paths)
+
+        # TODO - mtai @ davidmarin - As of 2016-02-26, the ONLY valid region is "global"
+        self['gcp_region'] = _DEFAULT_GCP_REGION
+
+        # Dataproc requires a master and >= 2 worker instances
+        # gce_num_instances refers ONLY to number of WORKER instances and does NOT include the required 1 instance for master
+        # In other words, minimum cluster size is 3 machines, 1 master and "gce_num_instances" workers
```

Comment:
This is similar enough to EMR that the Dataproc and EMR runners should use the same options and agree on what they mean. So, like, there should just be a `num_instances` option, and it should have to be at least 2 for Dataproc.

It's better to raise an exception than silently fix something like this.